tensorflow/contrib/lms/README.md

```diff
@@ -0,0 +1,179 @@
+      optimizer = tf.train.GradientDescentOptimizer(learning_rate=0.001)
+      train_op = optimizer.minimize(
+        loss=loss,
+	global_step=tf.train.get_global_step())
+```
+#### Step 2: define an LMSHook (LMSHook and LMS share the same set of parameters)
+```python
+# Hook for Large Model Support
+from tensorflow.contrib.lms import LMSHook
+lms_hook = LMSHook({'adam_optimizer'})
+```
+#### Step 3: add the LMSHook into Estimator's hook list
+```python
+mnist_classifier.train(
+      input_fn=train_input_fn,
+      steps=20000
+      hooks=[logging_hook, lms_hook])
+```
+
+For a working example of LMS integration with Estimator based training see:
+`examples/cnn_mnist_lms.py`
+which is an LMS enabled version of `https://github.com/tensorflow/tensorflow/blob/master/tensorflow/examples/tutorials/layers/cnn_mnist.py`.
+
+### [tf.keras](https://www.tensorflow.org/api_docs/python/tf/keras)-based training
+#### Step 1: define an LMSKerasCallback.
+```python
+from tensorflow.contrib.lms import LMSKerasCallback
+# LMSKerasCallback and LMS share a set of keyword arguments. Here we just
+# use the default options.
+lms_callback = LMSKerasCallback()
+```
+#### Step 2: pass the callback to the Keras `fit` or `fit_generator` function.
+```python
+model.fit_generator(generator=training_gen, callbacks=[lms_callback])
+```
+
+### Parameters for LMS/LMSHook/LMSKerasCallback
+#### Required parameters
+_graph_ :: the graph we will modify for LMS. This should be the graph of user-defined neural network. (not required in LMSHook and LMSKerasCallback)
+
+_optimizer_scopes_ :: scopes for the optimizers/solvers.
+
+#### Optional parameters
+_starting_scope_ :: Tensors that are reachable from the operations in this scope will be swapped for LMS. Set this to the scope of the first layer if we would like to modify the whole graph. Default `None`.
+
+_starting_op_names_ :: Tensors that are reachable from the operations with these names will be swapped for LMS. Default `None`.
+
+_excl_scopes_ :: a set of scopes for operations whose tensors will not be swapped out to the host. Default `empty`.
+
+_incl_scopes_ :: a set of scopes for operations whose tensors will be swapped out to the host. Default `empty`.
+
+_excl_types_ :: a set of types for operations whose tensors will not be swapped out to the host. Default `empty`.
+
+_incl_types_ :: a set of types for operations whose tensors will be swapped out to the host. Default `empty`.
+
+_n_tensors_ :: The number of tensors for LMS, counting from the `starting_scope`. To turn off LMS, set `n_tensors` to `0`. Default `-1` (all reachable tensors will be swapped for LMS).
+
+_lb_ :: Lowerbound value for LMS. A tensor will be swapped in during the backward phase at least `lb` nodes before it in the graph. Default `1`.
+
+_ub_ :: Upperbound value for LMS. Default `10000`.
+
+_fuse_swapins_ :: Fuse "close" swap-in operations into one operation. This may improve the performance. Default `False`.
```

Comment:
If it improves the performance, why `False` by default? Any side effect?

Reply:
Performance may be improved but the maximum batch size we are able to train is decreased because tensors are kept in GPU for a longer time to be reused by operations.